src/raftstore/store/store.rs

```diff
@@ -1456,6 +1617,257 @@ impl<T: Transport, C: PdClient> Store<T, C> {
+        let region_id = target_region.get_id();
+        if let Some(p) = self.region_peers.get(&region_id) {
+            let exist_epoch = p.region().get_region_epoch();
+            let expect_epoch = target_region.get_region_epoch();
+            // exist_epoch > expect_epoch
+            if util::is_epoch_stale(expect_epoch, exist_epoch) {
+                return Err(box_err!(
+                    "target region changed {:?} -> {:?}",
+                    target_region,
+                    p.region()
+                ));
+            }
+            // exist_epoch < expect_epoch
+            if util::is_epoch_stale(exist_epoch, expect_epoch) {
+                info!(
+                    "{} target region still not catch up: {:?} vs {:?}, skip.",
+                    tag,
+                    target_region,
+                    p.region()
+                );
+                return Ok(None);
+            }
+            return Ok(Some(p));
+        }
+
+        let state_key = keys::region_state_key(region_id);
+        let state: RegionLocalState = match self.kv_engine().get_msg_cf(CF_RAFT, &state_key) {
+            Err(e) => {
+                error!(
+                    "{} failed to load region state of {}, ignore: {}",
+                    tag, region_id, e
+                );
+                return Ok(None);
+            }
+            Ok(None) => {
+                info!(
+                    "{} seems to merge into a new replica of region {}, let's wait.",
+                    tag, region_id
+                );
+                return Ok(None);
+            }
+            Ok(Some(state)) => state,
+        };
+        if state.get_state() != PeerState::Tombstone {
+            info!("{} wait for region {} split.", tag, region_id);
+            return Ok(None);
+        }
+
+        let tombstone_region = state.get_region();
+        if tombstone_region.get_region_epoch().get_conf_ver()
+            < target_region.get_region_epoch().get_conf_ver()
+        {
+            info!(
+                "{} seems to merge into a new replica of region {}, let's wait.",
+                tag, region_id
+            );
+            return Ok(None);
+        }
+
+        Err(box_err!("region {} is destroyed", region_id))
+    }
+
+    fn schedule_merge(&mut self, region: &metapb::Region) -> Result<()> {
+        fail_point!("on_schedule_merge", |_| Ok(()));
+        let req = {
+            let peer = &self.region_peers[&region.get_id()];
+            let state = peer.pending_merge.as_ref().unwrap();
+            let expect_region = state.get_target();
+            let sibling_peer = match self.get_merge_peer(&peer.tag, expect_region)? {
+                // Wait till next round.
+                None => return Ok(()),
+                Some(p) => p,
+            };
+            if !sibling_peer.is_leader() {
+                info!("{} merge target peer is not leader, skip.", self.tag);
```

Comment:
is that means both the leader of the involved regions should be located on the same store?

Reply:
No, source follower can schedule merge too.